### PR TITLE
[#55] Search without backend endpoint

### LINKED ISSUE #55: Proof of concept of search without TermIt backend
In order to remove the TermIt API as a dependency, we need to start with a proof of concept of one API-less feature to ensure overall feasibility of the migration. The search endpoint / autosuggest box is a good candidate.

### PR DESCRIPTION
**Pull request by karelklima**
This is s proof of concept of how we can accomplish removing TermIt backend.

I reused the SPARQL query that TermIt used to fetch indexed data from database. The front end works slightly differently though. Changes:

- Both labels and descriptions are searchable
- Search includes only terms, vocabularies are omitted
- Total score is not summed for "words" - terms sharing the same label. Instead the maximum value is used, but realistically the score should be identical for all terms in most cases, since they're sharing a label.
- Instead of a label, the search snipped is provided, including highlighted search terms by the Lucene engine

This PR needs to be tested extensively before merging - we should compare search results of this solution and the original one.

@FilipKopecky Feel free to suggest changes as you see fit!

Resolves #55.

### CODE REVIEW

File: src/api/WordsAPI.ts [hunk 1]
```diff
@@ -2,6 +2,8 @@ import axios from "axios";
 import { useQuery } from "react-query";
 import _ from "lodash";
 import { API } from "../app/variables";
+import { firstValueFrom } from "rxjs";
```
Comment (karelklima) [inline src/api/WordsAPI.ts:5]: This function returns the first value from observable as Promise

File: src/api/WordsAPI.ts [hunk 2]
```diff
@@ -34,3 +36,65 @@ export const useSearch = (word: string | undefined) => {
     enabled: !!word,
   });
 };
+
+const getDirectSearchResult = async (word: string | undefined) => {
```
Comment (karelklima) [inline src/api/WordsAPI.ts:40]: I created a separate suite of helpers / hooks as the `useSearch` hook is used in multiple places (and I only wanted the main search bar to be affected).

File: src/api/WordsAPI.ts [hunk 3]
```diff
@@ -34,3 +36,65 @@ export const useSearch = (word: string | undefined) => {
...
+  // adds isWord flag when there are multiple
+  const result = _(data)
+    .groupBy("label")
+    .map((objs, key) => {
+      return {
+        label: key,
+        // include text with highlighted search string, should be the same for all objects with the same label
+        displayText:
+          objs[0].snippetField === "label" ? objs[0].snippetText : key,
+        total_score: _.maxBy(objs, "score")?.score,
```
Comment (karelklima) [inline src/api/WordsAPI.ts:78]: The `total_score` is max score, but I didn't want to change the data shape yet.

File: src/api/WordsAPI.ts [hunk 4]
```diff
@@ -34,3 +36,65 @@ export const useSearch = (word: string | undefined) => {
...
+        vocabularies: _.map(_.uniqBy(objs, "vocabulary"), "vocabulary"),
+      };
+    })
+    .orderBy("total_score", "desc")
+    .value();
+
+  return result;
+};
+
+export type DirectSearchResult = ReturnType<
```
Comment (karelklima) [inline src/api/WordsAPI.ts:90]: This is just a type helper to get a return type of the function above, minus the promise.

File: src/api/data/context.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  if (ct === "application/x-www-form-urlencoded") {
+    headers.set(
+      "Content-type",
+      "application/x-www-form-urlencoded; charset=UTF-8"
+    );
+  }
+  return fetch(resource, init);
+};
+
+export const context = createContext({
```
Comment (karelklima) [inline src/api/data/context.ts:22]: This is just the settings / context for the `@ldkit` library.

File: src/api/data/search.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    "@type": xsd.double,
+  },
+} as const;
+
+export type SearchInterface = SchemaInterface<typeof SearchSchema>;
+
+export const SearchResource = createResource(SearchSchema, context);
+
+// Search query utils
+const getWildcardString = (text: string) => {
```
Comment (karelklima) [inline src/api/data/search.ts:25]: `budova*` -> `budova*`
`budova` -> `budova budova*`
`jednopatrová budova` -> `jednopatrová budova budova*`

File: src/api/data/search.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+  if (text.endsWith("*")) {
+    return text;
+  }
+  const tokens = text.split(/\s+/);
+  const lastToken = text.split(/\s+/).slice(-1);
+  return `${tokens.join(" ")} ${lastToken}*`;
+};
+
+const getExactMatchString = (text: string) => {
```
Comment (karelklima) [inline src/api/data/search.ts:34]: `budova` -> `<em>budova</em>`

For exact matching of snippets in index.

File: src/api/data/search.ts [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+      ?entity a ${n(skos.Concept)} ;
+              ${n(skos.prefLabel)} ?label .
+    }
+    ?entity ${n(popisDat["je-pojmem-ze-slovníku"])} ?vocabulary .
+    ?entity ${n(lucene.score)} ?initScore ;
+            ${n(lucene.snippets)} _:s .
+    _:s ${n(lucene.snippetText)} ?snippetText ;
+        ${n(lucene.snippetField)} ?snippetField .
+    FILTER (lang(?label) = "cs")
+    BIND(IF(lcase(str(?snippetText)) = lcase(str(${l(
```
Comment (karelklima) [inline src/api/data/search.ts:70]: There is some scoring magic here - a score is calcullated based on whether the snippet comes from a definition or a label, and whether it is an exact match or not. This logic was imported from TermIt.

File: src/components/SearchBar.tsx [hunk 9]
```diff
@@ -49,18 +51,22 @@ interface SearchBarProps {
...
 
-  const onChangeHandler = (label: string | null) => {
-    const item = find<SearchItem>(data, { label: label ?? "" });
-    if (!item) {
-      history.push(`/search?label=${label}`);
-    } else if (item.isWord) {
-      history.push(`/disambiguation?label=${label}`);
+  const onChangeHandler = (option: DirectSearchResult | string | null) => {
+    if (typeof option === "string") {
+      history.push(`/search?label=${option}`);
```
Comment (karelklima) [inline src/components/SearchBar.tsx:59]: There is something in the search box, but it does not match a suggested option.

File: src/components/SearchBar.tsx [hunk 10]
```diff
@@ -109,7 +115,11 @@ const SearchBar: React.FC<SearchBarProps> = (props) => {
       noOptionsText="Nebyly nalezeny žádné výsledky"
       fullWidth
       freeSolo
-      options={data.map((item) => item.label)}
+      options={data}
+      getOptionLabel={(option: DirectSearchResult) => option.label}
+      renderOption={(option: DirectSearchResult) => (
+        <div dangerouslySetInnerHTML={{ __html: option.displayText }}></div>
```
Comment (karelklima) [inline src/components/SearchBar.tsx:121]: The snippets contain `<em>` tags which need to be preserved for the best UX.

File: src/components/SearchBar.tsx [hunk 11]
```diff
@@ -49,18 +51,22 @@ interface SearchBarProps {
...
   const history = useHistory();
 
-  const onChangeHandler = (label: string | null) => {
-    const item = find<SearchItem>(data, { label: label ?? "" });
-    if (!item) {
-      history.push(`/search?label=${label}`);
-    } else if (item.isWord) {
-      history.push(`/disambiguation?label=${label}`);
+  const onChangeHandler = (option: DirectSearchResult | string | null) => {
+    if (typeof option === "string") {
```
Comment (FilipKopecky) [inline src/components/SearchBar.tsx:58]: This doesn't seem to work as expected. When typing "Budova" it returns just the string and doesn't recognizte the fact that the "Budova" is present in the search as a word

Comment (FilipKopecky) [reply src/components/SearchBar.tsx:58]: I created a new commit which resolves this issue. I hope you don't mind

Comment (karelklima) [reply src/components/SearchBar.tsx:58]: Good catch. This is a UX issue though. What if someone actually wants to search for "Budova"?

**Review by FilipKopecky — APPROVED** [review]
The code looks great. I played around with it and it looks very promising. The search returns only valid terms and caching works as expected. I really like that the text is being highlighted in the suggestions. The possibility to search within the definitions is cool, maybe add some indication in the suggestions, which would show the user that the match is here because the search string is present in definition? But overall I really like it and I think we should try to get rid off the TermIt API everywhere.

**Review by psiotwo — COMMENTED** [review]
> - Both labels and descriptions are searchable

The /fts endpoint in TermIt should have been accessing both lucene indices -  
for labels and for definitions.

> Instead of a label, the search snipped is provided, including highlighted search terms by the Lucene engine

This seems a bit confusing to me - when searching, I would like to know whether I am picking a term, its definition, or a concept from the suggestion list. What if we either provide this information as a badge to the suggestion results or make all search results related to concepts, definitions of which are shown? This google-like view is however probably more suited for a search result list, not for a suggestion list.

**Comment by bindeali** [comment]
The functionality has two issues in my opinion, one major and one minor:
* I have encountered a situation where the autocomplete offers me a term that the application then cannot find ("Nebyly nalezeny žádné výsledky" page). A term I was able to consistently reproduce it with is "Subjekt informace".
image
image
* I don't like that the autocomplete isn't asynchronous. If I want to look up a term that is longer and I'm not quick enough with my typing, the operation interrupts the typing.

Also - I'm not sure if this belongs here or in a separate issue, but I personally don't like the layout of the "disambiguation" search page. Looking at "Fyzická osoba" here, the definitions visually blend together with the vocabulary name, making it harder to read than I think it needs to be. In addition, I think the three dots signifying that the entry is not the entire definition are a bit too small, at first I thought it saw something like a "additional options" button and tried to click on it. I would make it part of the definition's text "jiného státu...".
image

**Comment by FilipKopecky** [comment]
> * I have encountered a situation where the autocomplete offers me a term that the application then cannot find ("Nebyly nalezeny žádné výsledky" page). A term I was able to consistently reproduce it with is "Subjekt informace". I wouldn't consider the PR ready until this is resolved, the other comments can be addressed later, in my opinion.

You are absolutely right. Tbh I didn't notice this edge case. The search successfully ignores "obviously wrong" data (terms without vocabularies etc.) But this is caused by something else. I tried searching this term on TermIt (which doesn't ignore this result as well) and it shows this following message. 
TermIt err - Subjekt informace

I don't know what exactly causes this error. Maybe @psiotwo would have some idea. For now we can just send HEAD requests for each result and validate their state. Not the best solution, but it will do the trick.

**Comment by FilipKopecky** [comment]
> Also - I'm not sure if this belongs here or in a separate issue, but I personally don't like the layout of the "disambiguation" search page.

I agree. It becomes very apparent when these strings are long. I created a new issue #59 which mentions it.

**Comment by karelklima** [comment]
Thanks everyone for the feedback!

@FilipKopecky 

> maybe add some indication in the suggestions, which would show the user that the match is here because the search string is present in definition?

We could possibly include a definition snippet in these cases. Related to Petr's comments.

@psiotwo 

> The /fts endpoint in TermIt should have been accessing both lucene indices -
for labels and for definitions.

True. The app was filtering out the definition search results until now though - hence the change.

> This seems a bit confusing to me - when searching, I would like to know whether I am picking a term, its definition, or a concept from the suggestion list. What if we either provide this information as a badge to the suggestion results or make all search results related to concepts, definitions of which are shown?

I should have have been more clear here. The search returns concepts based on fulltext search in their labels and descriptions. The suggestion box always renders *labels*, with highlighted substring that matches the search criteria. If the match occurs in a description, then a label of the particular concept is provided, without any highlighting. Concepts that share a label are grouped together (resulting in the disambiguation page when clicked).

@bindeali 

> I have encountered a situation where the autocomplete offers me a term that the application then cannot find

Good catch, looks like a data and/or backend issue (not really related to this PR though). AFAIK there is a mismatch between the SPARQL endpoint that the search uses, and a TermIt backend that the rest of the app uses. This obviously needs to be resolved.

> I don't like that the autocomplete isn't asynchronous. If I want to look up a term that is longer and I'm not quick enough with my typing, the operation interrupts the typing.

Does it mean that you are not able to type anything when the data is reloading? If so then we need to fix that. I haven't run into this issue, will investigate.

**Comment by bindeali** [comment]
> > I don't like that the autocomplete isn't asynchronous. If I want to look up a term that is longer and I'm not quick enough with my typing, the operation interrupts the typing.
> 
> Does it mean that you are not able to type anything when the data is reloading? If so then we need to fix that. I haven't run into this issue, will investigate.

Actually, scratch that. I tested it again on both my fast PC and my  slow notebook and it seems to be fine. It may have been a combination of slow internet (eduroam...) and a (very) slow PC (with a battery life saving scheme enabled), as I tested it at school. There was noticeable delay then. On my slow PC on my home network, the delay is negligible to almost impercetible.

**Comment by karelklima** [comment]
> > > I don't like that the autocomplete isn't asynchronous. If I want to look up a term that is longer and I'm not quick enough with my typing, the operation interrupts the typing.
> > 
> > 
> > Does it mean that you are not able to type anything when the data is reloading? If so then we need to fix that. I haven't run into this issue, will investigate.
> 
> Actually, scratch that. I tested it again on both my fast PC and my slow notebook and it seems to be fine. It may have been a combination of slow internet (eduroam... the speed ranges from 1KB to 8MB per second, depending on location, # of connections, and what the network had for breakfast, I imagine) and a (very) slow PC (with a battery life saving scheme enabled), as on Monday I tested it at school. There was noticeable delay then. On my slow notebook on my home network, the delay is negligible to almost impercetible.

I think I was able to improve the frame drops quite a bit by removing unnecessary console logging in the underlying library. Some main thread blocking still occurs though, and it is related to parsing SPARQL queries and processing TTL responses by a Comunica library that I built the solution upon. If this is still not good enough then we'll need to move it to a worker, or perhaps replace the library with something less capable, but more performant.

**Comment by psiotwo** [comment]
> > * I have encountered a situation where the autocomplete offers me a term that the application then cannot find ("Nebyly nalezeny žádné výsledky" page). A term I was able to consistently reproduce it with is "Subjekt informace". I wouldn't consider the PR ready until this is resolved, the other comments can be addressed later, in my opinion.
> 
> You are absolutely right. Tbh I didn't notice this edge case. The search successfully ignores "obviously wrong" data (terms without vocabularies etc.) But this is caused by something else. I tried searching this term on TermIt (which doesn't ignore this result as well) and it shows this following message. TermIt err - Subjekt informace
> 
> I don't know what exactly causes this error. Maybe @psiotwo would have some idea.

Filip, which TermIt instance are You referring to now?

**Pull request merged**